internal/provider/client_test.go

```diff
@@ -22,12 +25,121 @@ func TestComputeURLPath(t *testing.T) {
 	var expected = "https://localhost:8043/api/v2/state/"
 	for _, tc := range testTable {
 		t.Run(tc.name, func(t *testing.T) {
-			client, err := NewClient(tc.url, nil, nil, true, 0)
-			if err != nil {
-				t.Fatalf(`Failed to create provider client %v`, err)
+			client := AAPClient{
+				HostURL:     tc.url,
+				Username:    nil,
+				Password:    nil,
+				httpClient:  nil,
+				ApiEndpoint: "",
 			}
 			result := client.computeURLPath(tc.path)
 			assert.Equal(t, result, expected, fmt.Sprintf("expected (%s), got (%s)", expected, result))
 		})
 	}
 }
+
+func assertTest[T int64 | string](t *testing.T, expectedErr string, err error, expectedValue T, result T) {
+	if len(expectedErr) > 0 {
+		if assert.Error(t, err) {
+			assert.Equal(t, expectedErr, err)
+		}
+	} else {
+		if assert.NoError(t, err) {
+			assert.Equal(t, result, expectedValue, fmt.Sprintf("Expected value (%v), got (%v)", expectedValue, result))
+		}
+	}
+}
+
+func TestGetKeyFromJson(t *testing.T) {
+	testTable := []struct {
+		name          string
+		keyName       string
+		expectedValue interface{}
+		err           string
+	}{
+		{name: "match string key", keyName: "current_version", expectedValue: "/api/controller/v2/", err: ""},
+		{name: "miss string key", keyName: "current", expectedValue: nil, err: "missing attribute 'current' from JSON response"},
+		{name: "match int64 key", keyName: "version", expectedValue: 1, err: ""},
+	}
+	jsonData := []byte(`{"current_version": "/api/controller/v2/", "version": 1}`)
+	for _, tc := range testTable {
+		t.Run(tc.name, func(t *testing.T) {
+			switch v := tc.expectedValue.(type) {
+			case string:
+				var res string
+				err := GetKeyFromJson[string](jsonData, tc.keyName, &res)
+				assertTest[string](t, tc.err, err, v, res)
+			case int64:
+				var res int64
+				err := GetKeyFromJson[int64](jsonData, tc.keyName, &res)
+				assertTest[int64](t, tc.err, err, v, res)
+			}
+		})
+	}
+}
+
+type mEndPointHttpClient struct {
```

Comment:
Particularly for testing our client code, I dislike the approach of mocking all the client functions and would prefer that we use the built-in httptest package as describe [here](https://medium.com/zus-health/mocking-outbound-http-requests-in-go-youre-probably-doing-it-wrong-60373a38d2aa). What do you think?

Reply:
updated